File: python/06_ownership.py
"""Lesson 06 — Ownership & Borrowing
Python side: demonstrating what Python does INSTEAD of ownership.
Rust's ownership has NO equivalent in Python — this file shows the contrast."""

import sys


def demo_python_no_ownership():
    """Python: multiple names can point to the same object.
    Rust: let s2 = s1; moves the value — s1 is DEAD."""
    s1 = "hello"
    s2 = s1  # Python: s2 now points to same string as s1
    print(f"Python — both alive after 'assign': s1={s1}, s2={s2}")
    # In Rust: let s2 = s1; println!("{s1}");  // ERROR: use of moved value


def demo_refcount():
    """Python tracks references at runtime.
    Rust tracks ownership at compile time — zero runtime cost."""
    a = [1, 2, 3]
    print(f"  refcount after a = [1,2,3]: {sys.getrefcount(a) - 1}")

    b = a
    print(f"  refcount after b = a: {sys.getrefcount(a) - 1}")

    c = b
    print(f"  refcount after c = b: {sys.getrefcount(a) - 1}")

    del b
    print(f"  refcount after del b: {sys.getrefcount(a) - 1}")
    # Rust: compile-time tracking, no refcount, no GC, no runtime cost


def demo_mutable_shared():
    """Python: shared references to mutable objects can cause surprises.
    Rust: prevents this at compile time."""
    inner = [1, 2, 3]
    outer = [inner, inner, inner]  # all three point to the SAME list
    inner.append(4)  # modifies all three at once
    print(f"Python shared mutation: {outer}")
    # Rust: cannot have multiple &mut references to the same Vec

    # In Rust, you'd need to clone explicitly:
    # let inner = vec![1, 2, 3];
    # let outer = vec![inner.clone(), inner.clone(), inner.clone()];


def demo_scope():
    """Python scoping — variables exist until function returns or del.
    Rust: values are DROPPED when the variable goes out of scope (RAII)."""
    x = "I exist"
    print(f"  before block: {x}")
    if True:
        y = "I'm inside the block"
        print(f"    inside block: {y}")
    # print(y)  # would error — y doesn't exist here
    print(f"  after block: {x}")
    # Rust: the same, but Rust also calls drop() when x goes out of scope
    # freeing resources immediately. No GC, no waiting, no cleanup() call.


def demo_python_clone():
    """Python: sometimes you explicitly copy to avoid shared mutation.
    Rust: MOVES by default, you .clone() if you want a copy."""
    original = [1, 2, 3]
    # Python: assignment is NOT a copy
    alias = original
    alias.append(4)
    print(f"  original after alias.append: {original}")  # mutated!

    # Python: explicit copy
    original2 = [1, 2, 3]
    copy = original2.copy()
    copy.append(4)
    print(f"  original after copy.append: {original2}")  # unchanged
    # Rust default is MOVE not share, so the 'alias problem' doesn't exist
    # Rust: let original = vec![1,2,3];
    #       let moved = original;         // original is dead
    #       // let alias = &original;     // ERROR: moved


def demo_tuple_vs_list():
    """Python: tuples are immutable, lists are mutable.
    Rust: mutability is a property of the BINDING, not the type."""
    # Python: tuple can't be changed
    t = (1, 2, 3)
    # t[0] = 99  # TypeError

    # Python: list can be changed
    lst = [1, 2, 3]
    lst[0] = 99

    # Rust: let t = (1, 2, 3);   → immutable binding
    #       let mut t = (1, 2, 3); → mutable binding
    # The tuple itself is the same type — the binding decides if it's mutable


if __name__ == "__main__":
    print("=== 1. No ownership in Python ===")
    demo_python_no_ownership()
    print("\n=== 2. Reference counting ===")
    demo_refcount()
    print("\n=== 3. Shared mutation ===")
    demo_mutable_shared()
    print("\n=== 4. Scoping ===")
    demo_scope()
    print("\n=== 5. Copy vs assign ===")
    demo_python_clone()
    print("\n=== 6. Mutability ===")
    demo_tuple_vs_list()

File: rust/06_ownership.rs
/// Lesson 06 — Ownership & Borrowing
/// Rust side: moves, clones, borrows, slices
///
/// This is THE Rust lesson. If you understand this, everything else clicks.
/// Python does nothing like this — ownership is 100% new.
///
/// The 3 rules of ownership:
///   1. Each value has exactly ONE owner
///   2. When the owner goes out of scope, the value is DROPPED
///   3. You can either: one &mut (mutable borrow) OR many & (immutable borrows)

/// ---- SECTION 1: Move semantics ----
/// In Python: let y = x;  → both y and x point to same object
/// In Rust:   let y = x;  → the value MOVES from x to y, x is dead

#[derive(Debug, Clone)]
struct House {
    name: String,
}

fn demo_move() {
    let h1 = House {
        name: String::from("Blue House"),
    };
    let h2 = h1; // ← MOVES: ownership transfers to h2

    // println!("{h1:?}");  // ❌ ERROR: use of moved value
    println!("move: owner is now h2 -> {h2:?}");

    // Python: h1 and h2 both point to the same house. Both alive.
    // Rust:   h1 is DEAD after move. Only h2 owns it.
    //         The compiler tracks this — no runtime cost.
}

/// ---- SECTION 2: Clone ----
/// If you want BOTH to be usable, you must .clone() explicitly.

fn demo_clone() {
    let h1 = House {
        name: String::from("Red House"),
    };
    let h2 = h1.clone(); // ← DEEP COPY

    println!("clone: h1 still alive -> {h1:?}");
    println!("clone: h2 is a copy -> {h2:?}");
    // Both alive. Two independent houses.
}

/// ---- SECTION 3: Copy types ----
/// Simple stack-only types (integers, bool, char, floats) implement Copy.
/// They don't move — they copy implicitly.

fn demo_copy() {
    let x: i32 = 42;
    let y: i32 = x; // ← COPY, not move (i32 implements Copy)

    println!("copy: x = {x}, y = {y}"); // both alive
    // Integers, bool, char, floats, tuples of Copy types → Copy
    // String, Vec, structs with non-Copy fields → MOVE by default
}

/// ---- SECTION 4: Borrowing (&) ----
/// Borrow = temporary access without ownership.
/// You give someone the key to look, but you still own the house.

fn look_at(house: &House) {
    // house is a REFERENCE (&House). We can READ but not move/destroy.
    println!("  visitor sees: {house:?}");
    // house is dropped here — but it's just a reference, the real value lives on
}

fn demo_borrow() {
    let house = House {
        name: String::from("Green House"),
    };

    look_at(&house); // ← borrow: house is NOT moved, we gave a reference
    println!("borrow: owner still has it -> {house:?}"); // ✅ still alive

    // look_at(house);  // ← would MOVE: house would be dead after
}

/// ---- SECTION 5: Mutable borrowing (&mut) ----
/// You give someone a renovation key — they can change things.
/// But only ONE renovation key can exist at a time.

fn renovate(house: &mut House) {
    house.name.push_str(" (renovated)");
}

fn demo_mut_borrow() {
    let mut house = House {
        name: String::from("White House"),
    };

    renovate(&mut house);
    println!("mut borrow: {house:?}"); // "White House (renovated)"

    // ---- The critical rule ----
    // You can have EITHER:
    //   - one &mut (mutable borrow)
    //   - OR many & (immutable borrows)
    // But NOT both at the same time.

    let ref1: &House = &house;
    let ref2: &House = &house;
    println!("  many &: {ref1:?} and {ref2:?}"); // ✅ many & is OK

    // let mut_ref: &mut House = &mut house;  // ❌ ERROR: can't borrow as mutable
    // println!("  read after mut? {ref1}");   //    because &ref1 and &ref2 are alive
}

/// ---- SECTION 6: The borrow checker in action ----

fn demo_borrow_checker() {
    let mut s = String::from("hello");

    // Rule: many & is fine
    let r1 = &s;
    let r2 = &s;
    println!("  many &: r1={r1}, r2={r2}");

    // The & references are no longer needed after last use above.
    // Now we can borrow mut:

    let r3 = &mut s;
    r3.push_str(" world");
    println!("  one &mut: {r3}");

    // What you CAN'T do:
    // let a = &s;        // immutable borrow
    // let b = &mut s;    // ❌ can't mix & and &mut
    // println!("{a} {b}");
}

/// ---- SECTION 7: Slices ----
/// A slice is a VIEW into a collection — it borrows without owning.

fn demo_slices() {
    let arr: [i32; 5] = [10, 20, 30, 40, 50];
    let slice: &[i32] = &arr[1..4]; // view of [20, 30, 40]

    println!("slices: arr={arr:?}");
    println!("slices: slice=&arr[1..4] => {slice:?}");
    // slice borrows arr — arr is still the owner, still alive
}

/// ---- SECTION 8: Dangling reference prevention ----
/// Rust prevents dangling references at compile time.

// fn demo_dangle() -> &String {  // ❌ ERROR: missing lifetime specifier
//     let s = String::from("hello");
//     &s  // ← s is dropped when function ends, reference would dangle
// }
// The compiler catches this. In Python, you'd get None or an error at runtime.

fn demo_no_dangle() -> String {
    let s = String::from("hello");
    s // ← MOVES ownership to the caller, no dangling
}

fn main() {
    println!("=== 1. Move semantics ===");
    demo_move();
    println!("\n=== 2. Clone ===");
    demo_clone();
    println!("\n=== 3. Copy types ===");
    demo_copy();
    println!("\n=== 4. Borrowing (&) ===");
    demo_borrow();
    println!("\n=== 5. Mutable borrowing (&mut) ===");
    demo_mut_borrow();
    println!("\n=== 6. Borrow checker ===");
    demo_borrow_checker();
    println!("\n=== 7. Slices ===");
    demo_slices();
    println!("\n=== 8. No dangling references ===");
    let s = demo_no_dangle();
    println!("  no dangle: {s}");
}
